tox.ini

```diff
@@ -0,0 +1,82 @@
+[tox]
+minversion = 3.14.0
+envlist =
+    fix_lint
+    py38
+    py37
+    py36
+    py35
+    py27
+    pypy
+    pypy3
+    cov
+    docs
+isolated_build = true
+skip_missing_interpreters = true
+
+[testenv]
+description = run tests with {basepython}
+setenv =
+    COVERAGE_FILE = {toxworkdir}/.coverage.{envname}
+    PIP_DISABLE_PIP_VERSION_CHECK=1
+    {py27,pypy}: PYTHONWARNINGS=ignore:DEPRECATION::pip._internal.cli.base_command
+extras = test
+commands =
+    pytest \
+    --cov {envsitepackagesdir}/build --cov-context=test \
+    --cov-report=html:{envtmpdir}/html \
+    --cov-report=xml:{toxworkdir}/coverage.{envname}.xml \
+    tests {posargs}
+
+[testenv:cov]
+description = combine coverage data
+deps =
+    coverage >= 5.0.1
+    diff-cover >= 3
+extras =
+skip_install = True
+passenv = DIFF_AGAINST
+setenv =
+    COVERAGE_FILE={toxworkdir}/.coverage
+commands =
+    python -m coverage combine
+    python -m coverage report --skip-covered --show-missing
+    python -m coverage xml -o {toxworkdir}/coverage.xml
+    python -m coverage html -d {toxworkdir}/htmlcov
+    python -m diff_cover.diff_cover_tool --compare-branch {env:DIFF_AGAINST:origin/master} {toxworkdir}/coverage.xml
+depends =
+    py38
+    py37
+    py36
+    py35
+    py27
+    pypy
+    pypy3
+parallel_show_output = True
+
+[testenv:docs]
+basepython = python3.8
+description = build documentation
+extras = doc
+passenv = *
+commands =
+    sphinx-build -d "{envtmpdir}/doctree" docs/source "{toxworkdir}/docs_out" --color -b html {posargs}
+    python -c 'import pathlib; print("documentation available under file://\{0\}".format(pathlib.Path(r"{toxworkdir}") / "docs_out" / "index.html"))'
```

Comment:
Why use the platform-dependent makefile that does not work on windows when you can use tox that works on all platforms? I want to actually remove that makefile, but thought it's out of scope of this PR.

Reply:
The makefile is usable on Windows with something like Cygwin os MinGW. If you want to remove it my answer would be to just rewrite it in python, instead of manually issuing commands in tox. This currently doesn't build the API reference.

Reply:
> The makefile is usable on Windows with something like Cygwin os MinGW

tox is usable without needing to install cryptic things like Cygwin or MingGW. You only need python, and you already have Python because how else would you run this project.

> If you want to remove it my answer would be to just rewrite it in python, instead of manually issuing commands in tox

Why would I do that? The whole point of tox (and nox too) is to hide away Makefile like functionalities in a cross-platform way, that's driven by config in case of tox, or the nox.py in case of nox. Writing that python file would be just rewriting tox in here. See my talk https://www.youtube.com/watch?v=SFqna5ilqig for more in-depth.

> This currently doesn't build the API reference.

True, but that's a follow-up ticket, has nothing to do with this PR.

Reply:
> tox is usable without needing to install cryptic things like Cygwin or MingGW. You only need python, and you already have Python because how else would you run this project.

I was not clear enough, sorry. What I was saying is that the makefile can be used on Windows, not that it would be usable in this scenario.

> Why would I do that? The whole point of tox (and nox too) is to hide away Makefile like functionalities in a cross-platform way, that's driven by config in case of tox, or the nox.py in case of nox. Writing that python file would be just rewriting tox in here. See my talk [youtube.com/watch?v=SFqna5ilqig](https://www.youtube.com/watch?v=SFqna5ilqig) for more in-depth.

Because I want to be able to build the docs without tox.

Reply:
> Because I want to be able to build the docs without tox.

This is not really possible in general. Someone needs to create a virtual environment, install the dependencies and then invoke the correct command to build the documentation. In theory, you can do this by invoking this commands yourself, but tox does all that for you automatically. Otherwise, who would ensure that all docs dependencies are satisfied, and they don't pollute the global namespace?

Reply:
Me as a distro packager. I want full reproducibility, so I take care of the environment and build the package, including the docs. This is not possible with a virtual env.

Usually what you want to do is make everything usable without a virtual env, and then use it inside a virtual env if you want. It is not your call if I get to build the docs, or do anything else, inside a virtual env or not. This is the kind of mentality that makes the job significantly harder for distro packagers, or any other kind of people that need more fine-tuned control.

Reply:
> Me as a distro packager. I want full reproducibility, so I take care of the environment and build the package, including the docs. This is not possible with a virtual env.

Why wouldn't be possible? tox will ensure all that given you provide a Python to use.

> It is not your call if I get to build the docs, or do anything else, inside a virtual env or not.

I agree with the fact that you run within a virtual environment or you bring your own Python requires flexibility, and should probably be an option. However, I don't agree with the fact that the project should not control how and what packages to the provision to perform its test suite. There's a better way though here, read https://discuss.python.org/t/proposal-for-tests-entry-point-in-pyproject-toml/2077 for a proposal made by Fedora on how to handle this in the long run. It also introduces https://github.com/fedora-python/tox-current-env#tox-current-env that allows running tox on a global python, rather than virtual environment where you can bring your own dependencies.

> This is the kind of mentality that makes the job significantly harder for distro packagers

Think of the Windows package managers now. Forcing a non-cross platform makefile (the fact that you can run it within cygwin does not matter, that project is a community best effort and not supported by MS itself) on them makes life significantly harder for them.

Reply:
> Why wouldn't be possible? tox will ensure all that given you provide a Python to use.

Reproducibility, I want to build a package today and be able to build it again in 5 years and get the *exact* same file. This is how you provide trust to the users, they do not have to rely on you saying that there is no backdoor, they can go ahead build the package themselves and get the same binary result. The language packaging (via PyPI, pip, ect.) is not there yet, but I hope it will be in the future. I would recommend you to take a look at https://reproducible-builds.org/. Installing via pip is a very big no-no for reproducibility.

If I test inside a virtual this also has absolutely no meaning what so ever for the distro, I am not testing if the package works properly with the distro packages, I am testing if the package works properly with the dependencies I fetched from PyPI and installed with pip.

> However, I don't agree with the fact that the project should not control how and what packages to the provision to perform its test suite.

Then we disagree. I have a *very* strong opinion here. I should be able to run everything in my environment, tox should just automate that. tox can make arbitrary choices on how to do it, but I should still be able to run pytest, mypy, sphinx, etc. manually.

> There's a better way though here, read [discuss.python.org/t/proposal-for-tests-entry-point-in-pyproject-toml/2077](https://discuss.python.org/t/proposal-for-tests-entry-point-in-pyproject-toml/2077) for a proposal made by Fedora on how to handle this in the long run. It also introduces [fedora-python/tox-current-env#tox-current-env](https://github.com/fedora-python/tox-current-env#tox-current-env) that allows running tox on a global python, rather than virtual environment where you can bring your own dependencies.

That is great. But still has some shortcommings.

> An attempt to run this with a Python version that doesn't match will fail (if tox is invoked from an Python 3.7 environment, any non 3.7 testenv will fail).

Pulling tox, the plugin and their dependencies also means that it takes more time to build packages, but I'll have to shut up and live that.

One of my main issues is that tox is not designed to run this way, it is entirely built around the idea of virtual environments. When you try to do something like this, you hit issues such as this one. I am not saying it is impossible to work around that, but it is very hard to do it properly. There is probably a reason this option is provided by a plugin and not by the upstream, right?

Right now, tox is not a tool I would want to use in distro packaging. Forcing its use a big no from me.

> Think of the Windows package managers now. Forcing a non-cross platform makefile (the fact that you can run it within cygwin does not matter, that project is a community best effort and not supported by MS itself) on them makes life significantly harder for them.

That is why I want to rewrite the makefile in Python, so that it is cross platform.

Reply:
> tox can make arbitrary choices on how to do it, but I should still be able to run pytest, mypy, sphinx, etc. manually.

tox wants to be the space where you can define what dependencies you request and how (more importantly what) you run. E.g. tox defines that use pytest to run the test suite and also contains the flags needed to pass in. The part that creates the environment to run that command into is configurable (and will be even further once tox 4 lands).

> That is why I want to rewrite the makefile in Python, so that it is cross platform.

tox wants to be this cross-platform tool though? Why rewrite tox when you can just use tox (which is also just pure python).

> One of my main issues is that tox is not designed to run this way, it is entirely built around the idea of virtual environments.

This is a false statement. Most people have this picture about tox mainly because that is what tox does by default, however, it's not true. We did not design tox with this enforcement in mind. By default, it uses virtual environments to achieve isolation, and pip to install dependencies. However, there's a reason both of these can be overridden entirely by plugins. This is not the only way though, through plugins you can customize both: e.g. use pacman to get python, translate PyPI dependencies to host dependencies, install them via pacman; and then ask tox to invoke the test commands, report the outcome. Each OS/distribution can create its own tox plugin and use its own tools to create a test environment, install dependencies, however the core logic of what are the dependencies, how to run them is left to drive for tox. 

> There is probably a reason this option is provided by a plugin and not by the upstream, right?

It's because the defaults are catering for 90% of people, what most people would want to use. tox does not want to dictate how to create environments/install dependencies for various distributions, so the behaviour is left up to the target dependency entirely that can be provided by plugins. This allows these plugins to evolve without needing input from tox maintainers.

Reply:
That is fair, but as it stands tox is undesirable for distro packagers like me. We should not enforce it's use like this.

Reply:
I'm not following why is less desirable than makefile.

Reply:
Because you can easily run the makefile without worrying about the environment. Right now for tox I need to use an external plugin which doesn't even work good enough for me.

> An attempt to run this with a Python version that doesn't match will fail (if tox is invoked from an Python 3.7 environment, any non 3.7 testenv will fail).

The main blocker is that the plugin does not simply skip if the environment is not compatible.

Reply:
How do all other projects using tox (this includes the majority of PyPa projects) manage? 

Anyways I give up. This project requires too much arguing for my side for adopting current best practices, so I resign from wanting to maintain it and focus on other projects.